processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -0,0 +1,457 @@
+  }
+
+  private void growIfPossible()
+  {
+    final int newBuckets = buckets * 2;
+    final int newMaxSize = maxSizeForBuckets(newBuckets);
+
+    if (newBuckets > (tableArenaSize - tableStart) / bucketSize) {
+      // Not enough space left to grow
+      return;
+    }
+
+    final int newTableStart = tableStart + tableBuffer.limit();
+
+    ByteBuffer newTableBuffer = buffer.duplicate();
+    newTableBuffer.position(newTableStart);
+    newTableBuffer.limit(newTableBuffer.position() + newBuckets * bucketSize);
+    newTableBuffer = newTableBuffer.slice();
+
+    int newSize = 0;
+
+    // Clear used bits of new table
+    for (int i = 0; i < newBuckets; i++) {
+      newTableBuffer.put(i * bucketSize, (byte) 0);
+    }
+
+    // Loop over old buckets and copy to new table
+    for (int oldBucket = 0; oldBucket < buckets; oldBucket++) {
+      if (isUsed(oldBucket)) {
+        final ByteBuffer entryBuffer = tableBuffer.duplicate();
+        entryBuffer.position(oldBucket * bucketSize);
+        entryBuffer.limit((oldBucket + 1) * bucketSize);
+
+        final ByteBuffer keyBuffer = entryBuffer.duplicate();
+        keyBuffer.position(entryBuffer.position() + HASH_SIZE);
+        keyBuffer.limit(keyBuffer.position() + keySize);
+
+        final int keyHash = entryBuffer.getInt(entryBuffer.position()) & 0x7fffffff;
+
+        final int newBucket = findBucket(newTableBuffer, newBuckets, bucketSize, true, keyBuffer, keySize, keyHash);
+        if (newBucket < 0) {
+          throw new ISE("WTF?! Couldn't find a bucket while resizing?!");
+        }
+        newTableBuffer.position(newBucket * bucketSize);
+        newTableBuffer.put(entryBuffer);
+
+        buffer.putInt(tableArenaSize + newSize * Ints.BYTES, newBucket * bucketSize);
+        newSize++;
+      }
+    }
+
+    buckets = newBuckets;
+    maxSize = newMaxSize;
+    tableBuffer = newTableBuffer;
+    tableStart = newTableStart;
+
+    if (size != newSize) {
+      throw new ISE("WTF?! size[%,d] != newSize[%,d] after resizing?!", size, maxSize);
+    }
+  }
+
+  private static int maxSizeForBuckets(int buckets)
+  {
+    return Math.max(1, (int) (buckets * MAX_LOAD_FACTOR));
+  }
+
+  /**
+   * Finds the bucket into which we should insert a key.
+   *
+   * @param keyBuffer key, must have exactly keySize bytes remaining
+   *
+   * @return bucket index for this key, or -1 if no bucket is available due to being full
+   */
+  private static int findBucket(
+      final ByteBuffer tableBuffer,
+      final int buckets,
+      final int bucketSize,
+      final boolean allowNewBucket,
+      final ByteBuffer keyBuffer,
+      final int keySize,
+      final int keyHash
+  )
+  {
+    Preconditions.checkArgument(
+        keySize == keyBuffer.remaining(),
+        "keyBuffer.remaining[%s] != keySize[%s]",
+        keyBuffer.remaining(),
+        keySize
+    );
+
+    final int startBucket = Math.abs(keyHash % buckets);
+    int bucket = startBucket;
+
+outer:
+    while (true) {
```

Comment:
which part would you put in a function?

I wanted to avoid recursion since the jvm can have trouble inlining that.